### ISSUE TITLE
Supercharging FENtastic!

### ISSUE THREAD

**Issue by ivarbrandt** [Tikipeter/repository.tikipeter#217]
Hey man!

I've decided on which parts of your skin I want to add to my skin. It's basically all of it lol. I really like the ability to set as many category widgets as I like. I also like the stacked widgets. These are really cool and after using them for awhile, I found them quite useful. There's one thing I would like to know before implementing the stacked widgets. This leads me to my first concern.

When scrolling through a stacked category widget in order to load the widget below, I noticed that the widget list loads instantly as soon as it is highlighted. I think this works best when scrolling one item at a time but when wanting to scroll over more than 1 item at a time, it subsequently loads all of the widget lists when they are highlighted even if it just for a split second. So when scrolling over say 10 items in the category widget, it will load all 10 widget lists before loading the one that you settled on. This causes longer loading times. I know that this behavior isn't the Helper's fault rather just the way the skin works. Do you know of a way that the loading of the widget list can be delayed by 5/10 or 7/10 of a second after being highlighted? This would make it so that you could quickly scroll through the category widget without loading unwanted lists in between. This is the biggest obstacle I see to implementing the stacked widgets.

My second question/concern is how do you generally recommend going about implementing all of these features? I looked at the changes you made in the Helper and it looks rather daunting. I know that the files that need to be changed in the skin are `Home.xml`, `Includes-Home.xml`, `Includes.xml`, and `SkinSettings.xml`. In the Helper, most of the changes are in `cpath_maker.py`. I'm just not sure how to start going about this. The first time that I implemented the Helper into the skin I just went around all willy-nilly breaking everything lol. This time I would like to take a bit more of a structured approach. Any advice would be greatly appreciated.

**Comment by Tikipeter** [comment]
Had a couple of days offline, sorry about that.

I read your post here earlier this morning, and I agree that a delay for the stacked widget to load would be nice. So I spent the morning getting that sorted out. It was a lot harder to make happen than I thought, but if you download the latest fenstruary and helper and remake your widgets, it should work.

The best method of attack for you would be to get the helper converted first with the new code. Replace all instances of fenstruary with fentastic, then replace any differently named includes as well. After that you can get started on the skin.

I'd be willing to take a shot at converting it for you. I just don't know how much of it you want to do yourself as a learning experience.

The helper should be done first though.

**Comment by ivarbrandt** [comment]
> Had a couple of days offline, sorry about that.

No need to apologize. I completely understand.

Thanks for adding the delay in. I was going to try to set that up in the skin rather than the helper. But you're a beast so of course you put it in the helper.

> The best method of attack for you would be to get the helper converted first with the new code.

Alright I'll get started on it soon. I want to learn as much as possible so I'm going to try to tackle it myself. If I get stuck I'll reach out to you on this issue so please don't close it yet.

Thanks so much for your input.

**Comment by Tikipeter** [comment]
I did one more update to fenstruary and the helper. I'll leave them alone now so you have a solid base to start with.

>I was going to try to set that up in the skin rather than the helper.

That would be better to be honest. So if you think of a way to do it within the skin, let me know. I might have a play with `AlarmClock` later and see if anything can be done with that.

**Comment by ivarbrandt** [comment]
> So if you think of a way to do it within the skin, let me know.

Will do. I'm going to see if I can use the new skin timers in Nexus. If not I'll figure out another way.

Were the only changes that you added in the helper in order to add the delay, the addition of the `service_functions.py` and `service.py` files? Or was there something else? I'm asking so I know what to remove in order to be able to test the delay that I'm going to add in the skin.

**Comment by ivarbrandt** [comment]
> If you just comment out the two calls for the two functions to run in services.py

This worked perfectly. I haven't been able to add the delay into the skin yet but I'm still working on it.

**Comment by ivarbrandt** [comment]
Hey! Hope you're having a good weekend. 

I have a little bit of a progress update for you. 

I was able to convert the helper and get my skin all set up to work with it. I didn't add in the delay that you added in the most recent versions because I wanted to find a way to add the delay into the skin. I ran into a problem though. When I tried to comment out the two functions in services.py on the helper in your skin, the stacked widgets stopped loading. It just loaded a normal category widget. So I left it out of my version of the helper for now.

Also, no luck on adding the delay in the skin. The biggest problem I'm running into is that when scrolling through the category widget, the focus doesn't change from that widget. So I'm not sure how to trigger the delay if there isn't a focus change.

**Comment by Tikipeter** [comment]
>When I tried to comment out the two functions in services.py on the helper in your skin, the stacked widgets stopped loading. It just loaded a normal category widget.

Yep, that will happen. The service function `widget_monitor` runs constantly, and when there is a focus on a category widget, it makes the `content_path` and `widget_header` values that the widget code uses to display the contents of the menu from the category widget above. It makes it as a window property.

The skin code then accesses those values by using for example `$INFO[Window(Home).Property(fenstruary.8012.path)]` and `$INFO[Window(Home).Property(fenstruary.8012.label)]`. Those are assigned within the `xmls.py` templates. i.e. `$INFO[Window(Home).Property(fenstruary.{cpath_list_id}.path)]`.

If not using this service function, the xml's would need to be made with

```
<param name="content_path" value="$INFO[Container({cpath_list_id}).ListItem.FolderPath]"/>\
<param name="widget_header" value="$INFO[Container({cpath_list_id}).ListItem.Label]"/>\
```

In the code above, `cpath_list_id` is the value of the parent widget. So `8011` for example.

The above is the correct way to do the stacked widget if not fiddling around with the delay.

So if you change the template for `stacked_media_xml_body` in `xmls.py` to include those 2 lines instead of what is currently there, then the stacked widgets should work. Just without the delay.

**Comment by Tikipeter** [comment]
>Also, no luck on adding the delay in the skin. The biggest problem I'm running into is that when scrolling through the category widget, the focus doesn't change from that widget. So I'm not sure how to trigger the delay if there isn't a focus change.

Yeah I know what you mean. It's disappointing.

**Comment by ivarbrandt** [comment]
> The above is the correct way to do the stacked widget if not fiddling around with the delay.

Thank you for the solid explanation. Got it working without the delay.

> Yeah I know what you mean. It's disappointing.

That it is. I will keep trying different ways to figure this out.

One more thing. I'm not sure if you've noticed this but when loading a category widget, scrolling down into the list, focusing a list item, then going back up into the category widget, and scrolling to a different category widget; if the item that you previously focused on exists in the next category widget, it will load the list with that item focused. This can make the focus start in the middle or the end of the list depending on where that item is. This is even more pronounced in my skin because I use a fixedlist for all of my widget items. I've attached a short clip that demonstrates this behavior.

https://user-images.githubusercontent.com/54084201/235380241-67cb41d0-4cc8-46bb-a76c-497487f57cdf.mp4

Is this behavior related to the helper or to the skin?

**Comment by Tikipeter** [comment]
Nasty Kodi related action I'm afraid.

Kodi wants to remember the last position you were in for each list, and return you to it. If the same listitem is in the next stacked list loaded, it will focus onto that. If not, it will usually focus on the position of where that listitem was in the previously loaded list.

This is another action that can be fixed quite easily in python code (we can re-set the list position to 0 everytime we load a new bottom stack list) but I couldn't see a way to do it using just the skin. Why hadn't I done it yet in the service code?..... I only just remembered the way to do it. :) In python, it's as easy as:

`window.getControl(int(list_id + '1')).selectItem(0)`

I'll be honest.... despite doing the changes I did to Estuary, I don't have a lot of knowledge remembered from 10 years ago when I made a very involved mod of confluence. I struggle each time I need to do something skin related. When I saw that Jurial's AH stacked widgets suffered from the same issue, I assumed it may not be possible to fix using the skinning engine.

I would say with the extra stuff you have done to FENtastic, you would likely have a better grasp of skinning between the two of us as it stands. :)

**Comment by ivarbrandt** [comment]
> If not, it will usually focus on the position of where that listitem was in the previously loaded list.

Kodi just likes to do stuff that wasn't asked of it lol.

> This is another action that can be fixed quite easily in python code

Awesome! I added the code in at the bottom of the widget monitor between `window.setProperty(path_prop % list_id, cpath_path)` and `try: del monitor`. Looks like it is working great!

> I struggle each time I need to do something skin related

I completely understand. I struggle every time I need to do something Helper related lol.

> When I saw that Jurial's AH stacked widgets suffered from the same issue, I assumed it may not be possible to fix using the skinning engine.

You were right. Now we have coolest skins on the block 😎 

> I would say with the extra stuff you have done to FENtastic, you would likely have a better grasp of skinning between the two of us as it stands. :)

Oh wow thank you for the kind words. If you need something added to Fenstruary let me know and I'll see what I can do to help.

Thank you so much for everything you've helped me with. I really appreciate it. 
I'm soo excited for this release of FENtastic! It's gonna be awesome!

**Comment by ivarbrandt** [comment]
I've run into a problem that I'm not sure how to fix. 

After I made sure everything was working correctly in my skin, it was time to add in the delay. 

1. So I created the service_functions.py and service.py files, added the functions `starting_widgets` and `widget_monitor` in service_functions.py, and called them in the service.py file. 
2. I also changed the `stacked_media_xml_body` include and swapped out

```
<param name="content_path" value="$INFO[Container({cpath_list_id}).ListItem.FolderPath]"/>\
<param name="widget_header" value="$INFO[Container({cpath_list_id}).ListItem.Label]"/>\
```
with

```
<param name="content_path" value="$INFO[Window(Home).Property(fentastic.{cpath_list_id}.path)]"/>\
<param name="widget_header" value="$INFO[Window(Home).Property(fentastic.{cpath_list_id}.label)]"/>\
```
3. I also changed `7` and `8` in this line: `list_id = '%s01%s' % ('7' if widget_type == 'movie' else '8', widget_number)` in the `starting_widgets` function to `19` and `22` as my control group for movies and shows are 19000 and 22000.

After all of these changes the stacked widgets should have worked but they weren't loading. It would load the category widget in the stacked widget but the actual poster widget or landscape widget was blank under the stacked category widget.

It was at this point that I realized that the changes to xmls.py would be calling the FENtastic skin itself in `$INFO[Window(Home).Property(FENtastic.{cpath_list_id}.path)]` and I knew that this could be a problem because I haven't be able to access skin properties in the past. For example, I've tried to access the version number of my skin with `$INFO[System.AddonVersion(skin.FENtastic)]` and it just doesn't work. I tried it on the Fen addon as well as your Fenstruary skin and it worked without a problem. The only thing I could think of that could be causing this is that Kodi doesn't like it when addon names begin with a capital letter.

At this point I'm thinking that the `addon id` in the skin's `addon.xml` is what is causing the issue. So I changed the `addon id` from `skin.FENtastic` to `skin.fentastic` and the helper from `script.FENtastic.helper` to `script.fentastic.helper`.

After doing this I made sure to change all instances of FENtastic to fentastic in the helper and the skin. Then I went to test the widget making features. All of the widgets worked except the stacked category widgets. So I decided to run a test to see if changing the `addon id` lets me access the skin properties and it worked. I added `$INFO[System.AddonVersion(skin.fentastic)]` to a different part of the skin and it returned the version number as expected. So I still had no idea why the stacked category widgets weren't working. 

I wanted to see if I had missed something during the addon id change so I commented out the `starting_widgets` and `widget_monitor` functions in the helper as well as reverted the xmls.py file. Lo and behold the stacked category widgets were working again!

At this point I think the problem lies somewhere in the delay, specifically the `starting_widgets` function. I compared the functions in my skin and yours and the only thing that is different is that is have `19` and `22` compared to your `7` and `8` in this line: `list_id = '%s01%s' % ('19' if widget_type == 'movie' else '22', widget_number)`. I cannot find any other differences and any other explanation for what is happening. 

What do you think may be happening?

**Comment by Tikipeter** [comment]
`$INFO[Window(Home).Property(FENtastic.{cpath_list_id}.path)]` is actually calling a home window property, not a skin property, so you were OK with that. But it is a good idea to change the addon id to all lowercase if it was causing you issues, as it would rear it's ugly head again in the future for sure.

In `Includes_Home`, make sure you have these 2 lines in the include `WidgetListCommonStacked`:
```
<onfocus>SetProperty(fenstruary_stacked_list_id,$PARAM[list_id],home)</onfocus>
<onunfocus>ClearProperty(fentastic_stacked_list_id,home)</onunfocus>
```

Also, check the `script-fenstruary-widget_movies.xml` and `script-fenstruary-widget_tvshows.xml` and see that the list ids for the stacked widgets are all what they are meant to be. If the stack widget is `19001`, then the widget under should be `190011`.

Also check the log for any errors in the `widget_monitor`, although that would give a notification that you would see.

If none of those solve it, zip up your skin and helper and I'll have a closer look....

**Comment by Tikipeter** [comment]
I saw another thing it might be...

Make sure you change `len(list_id) != 4` in the check within `widget_monitor` to reflect the length of your top level category widget `list_id`. Looks like it should be `5`.

**Comment by ivarbrandt** [comment]
> is actually calling a home window property, not a skin property, so you were OK with that. But it is a good idea to change the addon id to all lowercase if it was causing you issues, as it would rear it's ugly head again in the future for sure.

Good to know. And yeah I figured the same. I hope I don't have too many issues pushing it to my Kodi repo in the future.

Ok so I have checked everything you mentioned. I thought maybe the list ids were messed up because there is no longer a separate category widget include. But that seems to not be the case.

> Make sure you change len(list_id) != 4 in the check within widget_monitor to reflect the length of your top level category widget list_id. Looks like it should be 5.

I also tried this but it didn't work.

So here come the zips.

[skin.fentastic.zip](https://github.com/Tikipeter/repository.tikipeter/files/11377800/skin.fentastic.zip)

[script.fentastic.helper.zip](https://github.com/Tikipeter/repository.tikipeter/files/11377802/script.fentastic.helper.zip)

I removed the startup video so it would fit here on Github but I don't think that will give you any issues.

**Comment by Tikipeter** [comment]
Oh dear. I really should have told you to do this explicitly. I'm sorry mate.

In `addon.xml` of fentastic.helper, add this line:

`<extension point="xbmc.service" library="resources/lib/service.py" start="startup" />`

![Capture](https://user-images.githubusercontent.com/28930002/235798221-de536b5c-09ff-4bcc-836d-8451481f8dbc.JPG)

I'll look harder now through everything, bit I think that's the only issue, which means you did a fentastic(sorry) job converting things.

**Comment by ivarbrandt** [comment]
Ohh I had no idea the `addon.xml` had any changes done to it. That one line fixed it. Thank you so much!

> I'll look harder now through everything, bit I think that's the only issue

Ok let me know if you find anything.

> which means you did a fentastic(sorry) job converting things.

Thanks! And you just couldn't resist could you? 😆

**Comment by Tikipeter** [comment]
If you want a converted copy of fentastic using the version of widget monitoring I'm going to stick with moving forward.
There are no services running in this version (I didn't really like the fact I had to run a service for the widget monitoring).

[skin.fentastic-no_services.zip](https://github.com/Tikipeter/repository.tikipeter/files/11391603/skin.fentastic-no_services.zip)
[script.fentastic.helper-no_services.zip](https://github.com/Tikipeter/repository.tikipeter/files/11391604/script.fentastic.helper-no_services.zip)

**Comment by ivarbrandt** [comment]
Awesome thank you!

I was just about to release the update for FENtastic. Gonna have to add this in first for sure.

> The difference is the skin calls for the widget monitoring when onfocus of a stacked widget, and stops the process when onunfocus. So it's not running all the time waiting for a stacked widget.

That makes sense. What are the reasons you decided you don't want a service running? Does it impact performance?

**Comment by Tikipeter** [comment]
It's hard to say really, but it certainly doesn't help with performance.

Most addons will have some running processes in their services (Fen has 3 running continuously). It's just always better to not run a continual process if you can help it I feel. This one I just got rid of for the skin would check a few things every 0.75 seconds. As far as code goes that's not much effort, but I'm sure they all add up.

Good reason to disable any addons you might have installed as "backups" for when your "main" addon might go down. That'll stop them running service processes whilst just sitting there doing nothing.

**Comment by ivarbrandt** [comment]
> As far as code goes that's not much effort, but I'm sure they all add up.

This makes sense. I can definitely see this. I'm all about keeping things as speedy as possible.

> Good reason to disable any addons you might have installed as "backups" for when your "main" addon might go down. 

Funny you mention this. This is exactly what I used to do all the time back when I had Seren as a backup.

**Comment by ivarbrandt** [comment]
The latest version of FENtastic is now available from my repo.

Took quite awhile to get everything right as I really wasn't ready for how much trouble the addon id change would cause lol.

**Comment by Tikipeter** [comment]
Everything went according to plan. Updated this morning. :)

**Comment by ivarbrandt** [comment]
Awesome! Thanks so much for all of your help!

**Comment by ivarbrandt** [comment]
Hey! I hope this week finds you well.

I decided to make some more changes and I would like your opinion on something.

I wanted to find a way to fit more content on the home screen widgets. There were two ways I could do this. I could take off the InfoPanel so I could easily fit 6 poster and 4 landscape items per row. The other option was resizing the widgets. I like the InfoPanel so I decided to resize the widgets. This would also lead to a resizing across the InfoWall and Wall views.

So now there are 6 poster and 4 landscape items for the widgets instead of 4 and 3. I like how it turned out but I think that they might be too small, especially on smaller TVs. Also, the InfoWall and Wall views now have much more content per page but it is pretty tight in terms of space.

I've attached some screenshots for your viewing pleasure. I only included shows because I wanted to show the progress indicators as well. What do you think?

![Screenshot_20230509-121503](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/1d6d57ba-bc6d-4216-9e9b-6e4f81bc2f61)
![Screenshot_20230509-121545](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/e7f40244-edea-41f1-af8c-7d18a212c2a0)
![Screenshot_20230509-122347](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/9f10cf91-dba7-4820-bad9-67694396fec6)
![Screenshot_20230509-122410](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/fceaa493-8872-49ca-9fc5-140f1c745f05)
![Screenshot_20230509-124416](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/c2305183-7391-4c00-af46-3f9be86eac25)
![Screenshot_20230509-124428](https://github.com/Tikipeter/repository.tikipeter/assets/54084201/ea7826d3-c2c6-4c3a-b03f-9fff6141f20f)

**Comment by Tikipeter** [comment]
Looks good to me.

Probably most users would rather have more items on the screen I'd say.

You can only do what you think is best. :)

**Comment by les017** [comment]
@ivarbrant. Sure, since no one asked me, I'll chime in.

I can see preference for both, but can't be sure without seeing IRL. Would
a selection be possible or feasible?

Do love the skin!

**Comment by ivarbrandt** [comment]
Ok cool. I've been using it for the last few days and I'm really liking it.

> You can only do what you think is best. :)

Thank you for saying that.

**Comment by ivarbrandt** [comment]
@les017 

A choice would be possible but not feasible.

**Comment by TwilightMercy** [comment]
My opinion is don’t give up on InfoPanel! It looks amazing and makes the skin unique

**Comment by JBM05** [comment]
Where can get we this skin to try out?

**Comment by ivarbrandt** [comment]
That version of the skin hasn't been released yet. The changes you saw in the screenshots will be included in the next update.

**Comment by JBM05** [comment]
Ok, but where do I get the current version?  I don't know where your repo is.....thanks

**Comment by ivarbrandt** [comment]
Kodi File Manager Source: https://ivarbrandt.github.io/repository.ivarbrandt/

I ask everyone else to please direct future questions/comments about my skin to the Issues section of my repo:

I don't want to be blowing up Tikipeter with unnecessary notifications.

Thanks 😀

**Comment by ivarbrandt** [comment]
Hello Peter! How are you doing?

I was thinking about a few things wanted to add to the skin and I have a few questions about how you would go about it.

1. Extra Sections - I would like to add an additional 5 home main menu items and their respective widgets. Do you think I should do this dynamically in the code or would it be easier to just add them individually the way I have the Movies and Shows sections?

2. Ratings - I would like to add ratings to the InfoPanel. Can the ratings that are available through Fen be displayed similar to the watched progress? This would be simplest way for me to implement this.

3. Stacked widget delay - I would like to add a customizable delay to the stacked widgets as not everyone scrolls at the same speed. How do I set a variable in the helper and then call it in the skin?

P.S. I found a couple of the bugs that you mentioned you left in there for me and I have fixed them 😄 
There was one that put the 10th widget after the second widget on the home screen and another where if the 10th widget was a stacked widget it wouldn't load on startup.

**Comment by Tikipeter** [comment]
Extra Sections:
Would be easier adding directly into the skin code. Same for the widgets. What extra 5 list items are you hoping to add?

Ratings:
Trickier than it might first appear. You couldn't really use Fen's ratings, as they are only made when the Extras window is opened. The python code to get the ratings is fairly straight forward, and could be copied over largely to your helper. But the skin would have to trigger invoking the code for each list item.

Stacked widget delay:
This has been implemented in the latest 1 or 2 versions of fenstruary helper. It requires a change to the skinsettings.xml of course for the new setting(s), and then a change to widget_utils/widget_monitor to fetch the setting and deal with the delay.

You can't simply change the `monitor.waitForAbort()` time to a larger time, as that wait time is running continually, so when a change in focus comes it might be 0.2 seconds into a 3 second delay, or it might be 2.7 seconds into it. So the delay needs to start fresh when a new item is focused. That meant adding some new code into that function.

The setting I used is called `category_widget_delay`. I used a button setting so you can go through from 0.50 seconds to 5.00 seconds. I used the button setting because I could copy it from where it was already being used for similar. Wouldn't have been my first choice. :) I also added a setting if you wanted the stacked widget's label to show the delay as it happened (`category_widget_display_delay`).

Access the new skin setting(s) in the python code with `xbmc.getInfoLabel('Skin.String([SKIN_SETTING])')`, e.g `xbmc.getInfoLabel('Skin.String(category_widget_delay)')`

**Comment by ivarbrandt** [comment]
> Would be easier adding directly into the skin code. Same for the widgets. What extra 5 list items are you hoping to add?

I figured it would be. Well the extra sections aren't really for me. I've had a few people ask me for extra sections to set up Anime, Sports, Live TV, etc. I didn't really understand why as all I need is the Movies and Shows section. Then my sister mentioned to me that she would also like the ability to customize some extra sections. She helped me understand how it would be helpful for her and others so I decided to give it a shot. Plus coding it would be a fun challenge. So the 5 section names would be customizable by the user, similar to setting the widget labels. 

The only thing I didn't want is to lose the focus of the skin being purely centered around Fen. Adding sections probably wouldn't do that but that was my biggest concern.

> The python code to get the ratings is fairly straight forward, and could be copied over largely to your helper. But the skin would have to trigger invoking the code for each list item.

Ok, I might need your help with that. Before I start this I want to know, would calling the ratings for every listitem be inefficient or result in too many API calls?

> This has been implemented in the latest 1 or 2 versions of fenstruary helper. It requires a change to the skinsettings.xml of course for the new setting(s), and then a change to widget_utils/widget_monitor to fetch the setting and deal with the delay.

Awesome! Thank you for all of the info. I'll take a peek at the code as soon as I get a chance.

> I used the button setting because I could copy it from where it was already being used for similar. Wouldn't have been my first choice. :)

Lol I completely understand. I'll probably use a dialog window though for simplicity.

**Comment by Tikipeter** [comment]
>Plus coding it would be a fun challenge.

Pretty big job. At least if you're also going to offer custom widgets for the custom menu items.

>would calling the ratings for every listitem be inefficient or result in too many API calls?

Yes and yes. Whilst it won't slow the skin down, until the values are cached, each focused listitem would need a call to OMDb for the rating values. And they'd need to be refreshed periodically, otherwise users will whinge they aren't up to date enough. Think long and hard on adding it.

I also don't know how to incorporate it into the skin, but I can look what other skins are doing with regard to it and help out if you decide to go that route.

>Lol I completely understand. I'll probably use a dialog window though for simplicity.

Then I'll probably copy it. :)

I'm kinda treating fenstruary as a testing ground for things you might like to add to FENtastic. I saw that issue you had open on your github and thought I'd implement that delay and see if you mentioned it later. I don't think there's anyone still using Fenstruary. Hopefully not for their sake.

**Comment by ivarbrandt** [comment]
> Pretty big job.

I was looking at the code and yeahhh this would be a pretty big job so I think I'm gonna hold off on the extra sections for a bit.

> Think long and hard on adding it.

I will. I didn't think about the periodic refreshing. Thanks for the heads up.

> I'm kinda treating fenstruary as a testing ground for things you might like to add to FENtastic

Ah I see. Thank you! I'll keep a closer eye on it then. Could you please upload Fenstruary so I can see all of the files and changes in Github? This way I can more easily see the changes you've made to it rather than hunting down the changes in each release. Or could I get read access to your test repo? Idk what would be easier.

> I don't think there's anyone still using Fenstruary. Hopefully not for their sake.

Lmao why do you say that?

**Comment by TwilightMercy** [comment]
> > Pretty big job.
> 
> I was looking at the code and yeahhh this would be a pretty big job so I think I'm gonna hold off on the extra sections for a bit.
> 
> > Think long and hard on adding it.
> 
> I will. I didn't think about the periodic refreshing. Thanks for the heads up.
> 
> > I'm kinda treating fenstruary as a testing ground for things you might like to add to FENtastic
> 
> Ah I see. Thank you! I'll keep a closer eye on it then. Could you please upload Fenstruary so I can see all of the files and changes in Github? This way I can more easily see the changes you've made to it rather than hunting down the changes in each release. Or could I get read access to your test repo? Idk what would be easier.
> 
> > I don't think there's anyone still using Fenstruary. Hopefully not for their sake.
> 
> Lmao why do you say that?

My small Tip: I use "WinMerge" software (Windows PC) to import two .zip files (two addon versions) and it shows me the exact changes between them in the code

**Comment by ivarbrandt** [comment]
@TwilightMercy 

That's cool. I'll start using that. Thanks for the suggestion.

**Comment by ivarbrandt** [comment]
Hey! I have a question for you Peter. 

1. Do you know of a way to reload a single XML file in Kodi without calling `ReloadSkin()`? I ask because calling reload skin is fine for the main menu and widgets usage and configuration. But when implemented with the search results window, it can be a bit much depending on how much you're searching. Also, calling reload skin, introduces some jankiness to the UI that I would prefer to avoid. I did my best eliminate said jankiness in `1.0.12e` but if I can avoid the skin reload altogether that'd be awesome.

**Comment by Tikipeter** [comment]
I don't think that's possible. To simply reload one xml file.

What do you need to achieve by doing that? There might be another way to do it.

**Comment by ivarbrandt** [comment]
I want to be able to update the XML where I keep the search history. Right now the skin reloads if I add something to the database, clear the whole database, or open the search results window directly. I would like to be able to skip the skin reload for when I add something to the db. I don't think the other ones are doable as I am setting certain skin settings before and after skin reloads.

**Comment by Tikipeter** [comment]
Past my pay grade I'm afraid. :) It seems to work very well how it is now though (1.0.13)

BTW, next Fen will have the ability to both search and browse Trakt lists without a Trakt account. It is not needed for that, but I had it as a prerequisite. That will get rid of the Fen error in your search feature.

**Comment by ivarbrandt** [comment]
> Past my pay grade I'm afraid. :) It seems to work very well how it is now though (1.0.13)

I doubt there's much past your pay grade. You're an absolute beast of a programmer. :)
I'm glad it's not too noticeable. There's quite a few skin reloads in between actions lol.

> That will get rid of the Fen error in your search feature.

Sweeet thanks! I noticed those errors and I was able to mitigate them by setting the search path query to `"none"` instead of an empty string. I did this by creating a new string specifically for the Trakt queries called `SearchInputTraktEncoded`.
Great to know I won't need to do that when the latest Fen drops. It'll save a skin reload or two.

**Comment by ivarbrandt** [comment]
Hey Peter! Hope you’re doing well!

Before I start what is going to be a lengthy post I want to express that I completely understand and respect your current stance on no requests. So please, with zero pressure, get back to me when you are able and willing.  

I have added ratings to FENtastic and I would like your opinion on a few issues that I’m having. Also, I would like to know what you think about the overall implementation. I decided to not use a service based on previous advice I’ve received from you, but it looks like I might have to do just that. You can find the latest changes in `1.0.18c`. To set up the ratings, enter your OMDB API Key in Skin settings. 

The way I have the logic set up is so that when an item is focused it grabs the IMDB ID and then checks to see if the ID has a match for an entry in the database. If it does then it checks to see if the last time the rating was updated was more than 7 days ago. If it is younger than 7 days, it will fetch the entry from the database instead of hitting the OMDb API. If it is older than 7 days, it will hit the API and then store the updated ratings and timestamp. Then the ratings are stored in window properties which are accessed by `Window.Property()` infolabels and stored in variables. Those are then called in the InfoPanel and widget labels.

I’ve been testing it for awhile and there are a few consistent issues that I keep running into:

1. The ratings only work when scrolling from widget to widget but not when scrolling within a widget. I think this is because I’m not passing in a `list_id` through the `test_api` function in the `<onfocus>` tag that I’m using in `Includes_Home.xml`. This is something I haven’t figured out how to do.

2. I’m constantly getting SQLite Integrity errors. I think this is happening because when you scroll fast through items, it tries to write multiple entries to the database at the same time and this has caused incorrect ratings to be stored for different items as they sometimes overlap.

3. The ratings aren’t working within the Fen add-on where the views are. There isn’t an `<onfocus>` tag to trigger the function to fetch the ratings so idk how to get this working except like I said, implement a service.

4. Sometimes, like the sqlite integrity errors, I’ll get an error that has to do with the `datetime` import in python. It says that the `datetime` object is non callable. Restarting Kodi fixes this issue but it rears its ugly head randomly from time to time.

5. The ratings are kind of slow even if they are fetched from the database. I’m not sure if this can be improved upon by implementing the fetching of ratings as a service.

I can provide logs of the error related issues if you need them.

I’d really appreciate it if you could take a look whenever you get a chance. I remember you telling me it’s trickier than it would first seem to implement ratings and holy shit you were right.

**Comment by Tikipeter** [comment]
@ivarbrandt 

I'm really sorry.
I didn't see this until reading Umbrella dev's Issue with you on your github. I don't have notifications to my email turned on so I never know when someone has replied to an issue without checking.
I'll open this back up and will be able to help from tomorrow onwards hopefully.

**Comment by Tikipeter** [comment]
I spent a few hours last week experimenting, trying to work out why Fen crashes FENtastic after playback and returning to home menu. Because it doesn't do that on Fenstruary or most other skins since the languageinvoker fix was implemented a while ago.

I came up blank, but I did manage to see that the invoker fix did not even get a chance to run before the crashing, which leads me to believe it is in some way also related to some kind of inability for Kodi to find a control to focus on when first entering the homepage. Kodi will often crash if it is trying to focus on a control that doesn't exist.

But again, poking around the code I couldn't see why this would be the case.

**Comment by ivarbrandt** [comment]
> I didn't see this until reading Umbrella dev's Issue with you on your github.

Not a problem man. Great to hear from you!

> I'll open this back up and will be able to help from tomorrow onwards hopefully.

Awesome thank you!

> it is in some way also related to some kind of inability for Kodi to find a control to focus on when first entering the homepage.

I'm gonna reply to this here instead of on Umbrella dev's issue because they are different issues. I'll reply to the ratings issue over there.

I think I know what may be happening. The InfoPanel has a conditional visibility tag where it is shown when the focused item is either a movie, tvshow, season, or episode.

`<visible>String.IsEqual(ListItem.DBtype,movie) | String.IsEqual(ListItem.DBtype,tvshow) | String.IsEqual(ListItem.DBtype,season) | String.IsEqual(ListItem.DBtype,episode)</visible>`

If this results to false, then the main menu panel is shown instead. Here is the visibility tag for the main menu panel.

`<visible allowhiddenfocus="true">![String.IsEqual(ListItem.DBtype,movie) | String.IsEqual(ListItem.DBtype,tvshow) | String.IsEqual(ListItem.DBtype,season) | String.IsEqual(ListItem.DBtype,episode)]</visible>`

The reason the skin can switch back and forth between showing the main menu panel/hiding the InfoPanel and vice versa is because `allowhiddenfocus` is set to true for the main menu panel. Otherwise it wouldn't work.

Because the Infopanel is visible every time any type of content is played from a widget on the homescreen, then this means that the main menu panel is hidden. Therefore when playback ends, it goes back to the homescreen and tries to locate the default control, `<defaultcontrol>9000</defaultcontrol>` which is the main menu panel with an ID of 9000, but it can't find it because it is hidden. Then it crashes.

How to fix this though I am not sure..

**Comment by Tikipeter** [comment]
I've added most things into Fentastic now from your latest changes. Excluding OMDb ratings.

Fen is not causing a Kodi shutdown even now with Fentastic.

The only major change I made that I thought MIGHT have had an impact is using different logic for the display of the InfoPanel on the home screen.

Unfortunately it has not fixed the issue.

I'm at a loss as to what the cause is now. :-(

**Comment by ivarbrandt** [comment]
That is so strange. I was almost sure it was the main menu being hidden when returning to the homescreen.

The only other thing I can think of is there is something wrong with one of the widgets. And that would stem from the views. I'm gonna try having all poster widgets on the homescreen and see if it happens. Then I'll try the rest to see if there's one specific widget type that causes it to crash.

**Comment by ivarbrandt** [comment]
I re-read your comment now that I'm more awake and I'm a bit confused.

> Fen is not causing a Kodi shutdown even now with Fentastic.

Here are you saying there isn't any crashing anymore?

> Unfortunately it has not fixed the issue.

And here you're saying it it still crashing?

**Comment by Tikipeter** [comment]
Yeah I wrote the wrong skin. Got confused with the names. :-)

Fenstruary is still not crashing, Fentastic IS still crashing.

Bad news only.

**Comment by ivarbrandt** [comment]
> Fenstruary is still not crashing, Fentastic IS still crashing.

But why?? 😭 

I'm gonna take a deeper dive into it later. It literally makes no sense. I loaded up Fenstruary and I thought I was in FENtastic. Looks the EXACT same lol.

**Comment by ivarbrandt** [comment]
I changed/rearranged a few lines in the `WidgetListCommon` include in `Includes_Home.xml` and so far it hasn't crashed. I tried watching a movie for 15 to 20 mins at a time and then going back to the homescreen and it was fine. I also tried staying in other windows for awhile and then returning to the homescreen and it was fine. 

So idk if it's fixed or not. Because it doesn't make sense that the lines I changed would fix it. Could you try it and lemme know if it works for you?

I changed `WidgetListCommon` to this like it is in Fenstruary:

```
<include name="WidgetListCommon">
	<scrolltime tween="cubic" easing="out">500</scrolltime>
	<orientation>horizontal</orientation>
	<onleft>9000</onleft>
	<onright>$PARAM[list_id]</onright>
	<onback>9000</onback>
	<onback>SetFocus($PARAM[list_id],0,absolute)</onback>
	<onfocus>Control.Move($PARAM[list_id],-$INFO[Container($PARAM[list_id]).Position])</onfocus>
	<onfocus>Control.Move($PARAM[list_id],$INFO[Window(home).Property(listposition)])</onfocus>
	<onunfocus>SetProperty(listposition,$INFO[Container($PARAM[list_id]).Position],home)</onunfocus>
</include>
```

**Comment by Tikipeter** [comment]
I'll try it now and let you know. :)

**Comment by Tikipeter** [comment]
Still crashing on Windows. :( :( :(

**Comment by ivarbrandt** [comment]
Aw man. Ok how are you testing it? So that I can try to make it crash the same way.

**Comment by Tikipeter** [comment]
Navigate into Fen, play a movie (even for a few seconds). Then navigate back to home screen. For me it crashes every time.

**Comment by ivarbrandt** [comment]
Whoa I tried that once and then Fen threw an error and it instantly crashed. No problems playing things directly from the homescreen widgets though.

**Comment by ivarbrandt** [comment]
When you go into Fen, play something, then go to the homescreen, are there stacked widgets in the section you return to? If there is, try removing the stacked widgets then try again. 

I'm haven't been able to get an error again but I have noticed that when I go back to the homescreen the stacked widgets don't load correctly. Either the category widget will be empty, the main widget will be empty or there will just be a huge space on the screen where it used to be.

This is not happening on Fenstruary at all. The stacked widgets are loading correctly right after returning to the homescreen.

**Comment by SoulReaverFan** [comment]
I don’t know if this has anything to do with the difference in terms of the crashes, but I noticed that the TikiPeter test repo only updated the skin and not the helper addon. But maybe locally he is using all of the latest stuff. Just wanted to mention it just in case that could be a reason :) hope you guys figure it out. Have a good one and thanks for all the great work you do! 👍

**Comment by ivarbrandt** [comment]
I went down the path of looking at the stacked widgets. The includes looked the same in the skin XMLs so that led me to the Helper. The `starting_widgets` functions are the same and so is the` files_get_directory` function that it calls. The functions that weren't completely the same were `fetch_current_cpaths`. Specifically, the sorting was different. So I got something else for you to try Peter. I think the widgets are expecting lexicographical ordering but the Helper was sorting numerically. Change this line: 

`results.sort(key=lambda k: int(k[0].split(".")[-1]))`

 to 

`results.sort(key=lambda k: k[0])`

in the `fetch_current_cpaths` function in the FENtastic Helper. This fixed my stacked widgets breaking after playback. I think that maybe for you they don't break, the whole thing just crashes.

**Comment by Tikipeter** [comment]
Lexi what now?!

>EDIT: Nevermind. Still not fixed

Oh balls. Just got home from work. Was all excited to give it a try.

**Comment by ivarbrandt** [comment]
Haha it means string-based ordering.

I would still try it. I'm not on a Windows machine so maybe it'll be different for you.

It worked fine in the morning for me then I tested it again in the afternoon and the stacked widgets were intermittently broken lol.

**Comment by ivarbrandt** [comment]
I finally put Kodi on my Windows machine and FENtastic crashed every single time after playback. This was a good thing because it allowed me to actually test it as it barely crashed on my Shield.

I don't think I found what's actually causing the crashes but I may know why Fenstruary doesn't crash. 

I swapped out FENtastic's `Home.xml` for Fenstruary's and even after that it was still crashing. Then I went into Skin settings and toggled the Videos section to on. Suddenly, FENtastic stopped crashing. I thought I got lucky so I tried it again and again, no crash. Then I toggled the Videos section again to off and the crashing came back. I was absolutely flabbergasted.

So I loaded up Fenstruary to see if the same thing would happen. I noticed the Videos section was toggled to on in Fenstruary so I went into Skin settings and toggled it off. I started playback, ended it, and returned to the homescreen. 1 second later Fenstruary crashed. I tried it another ten times with the Videos section toggled to off and it crashed every single time. I toggled the Videos section back on and it completely stopped crashing.

I have no clue why this is happening. Can you please test this out? I need to know whether or not I'm crazy or my computer is just buggin hard.

**Comment by Tikipeter** [comment]
Yep, that's what's happening on my setup as well.

Why? Not really any idea.

I've tried a fair few things just now to work it out, but not really any reason behind it yet.

I'll try a couple of things in Fen too that could be related. Doubtful though.

**Comment by ivarbrandt** [comment]
I fixed it! Whoohoo!

In FENtastic, move the entire control group with an id=8000 on to the top above the movie widgets control. This fixed the crashing for me. Try it out and lemme know. 

Also let me know if I wasn't too clear as I'm just so excited it's fixed so I'm just running around all over my house celebrating 😄

**Comment by Tikipeter** [comment]
>I'm literally just running around all over my house celebrating because I'm so happy it's fixed! 😄

Hey, good for you man!

I tried having an "invisible" (placed at left -1000) addons widget list right at the top of the menu list, but it didn't fix it. But with more fiddling around that might be the way to go. Otherwise users will NEED to keep that addons menu item in the menu list as you know.

At least we know what is happening somewhat now though..... with the information that widgets load top to bottom on the home screen. Which makes sense as the xml is read top to bottom.

**Comment by ivarbrandt** [comment]
>  tried having an "invisible" (placed at left -1000) addons widget list right at the top of the menu list

I was thinking of doing something similar. Just place a copy of the addons control group at the very top with a different id and zero visibility conditions so that it's always showing, then like you said move it all the way off screen to the left. This way even if users toggle the addons section to off, there will always be a "hidden" addons section that will prevent crashing.

> with the information that widgets load top to bottom on the home screen

Yep, this seems to be the case. For some reason it needs to load the addons section where Fen is located before it loads the widgets. Otherwise RIP.
